packages/react-charting/src/components/SankeyChart/SankeyChart.base.tsx

```diff
@@ -350,15 +363,261 @@ export function preRenderLayout(
+ * provide color to  individual node.
+ */
+function assignNodeColors(
+  nodes: SNode[],
+  colorsForNodes: string[] | undefined,
+  borderColorsForNodes: string[] | undefined,
+) {
+  let colors: string[];
+  let borders: string[];
+  if (colorsForNodes && borderColorsForNodes) {
+    colors = colorsForNodes;
+    borders = borderColorsForNodes;
+  } else {
+    colors = DEFAULT_NODE_COLORS.map(color => color.fillColor);
+    borders = DEFAULT_NODE_COLORS.map(color => color.borderColor);
+  }
+  let currentIndex = 0;
+  nodes.forEach((node: SNode) => {
+    if (!node.color && !node.borderColor) {
+      node.color = colors[currentIndex];
+      node.borderColor = borders[currentIndex];
+    } else if (node.color && !node.borderColor) {
+      node.borderColor = '#757575';
+    } else if (node.borderColor && !node.color) {
+      node.color = '#F5F5F5';
+    }
+    currentIndex = (currentIndex + 1) % colors.length;
+  });
+}
+
+/**
+ * @param text is the text which we are trying to truncate
+ * @param rectangleWidth is the width of the rectangle which will contain the text
+ * @param padding is the space we need to leave between the rect lines and other text
+ * @param nodeWeight is the text if present needs to be accomodate in the same line as text.
+ * @returns the truncated text , if truncated given the above parameters.
+ */
+function truncateText(text: string, rectangleWidth: number, padding: number) {
+  // NOTE: This method is the most-expensive in terms of rerendering components.
+  const textLengthForNodeName = rectangleWidth - padding; // This can likely be computed once and passed in.
+  // The following `select` statement injects a `tempText` element into the DOM. This injection
+  // (and subsequent removal) is causing a layout recalculation. This is a performance issue.
+  // Note that this code will always inject a `tempText` element, but doesn't always remove it. This is a bug.
+  const tspan = select('.nodeName')
+    .append('text')
+    .attr('class', 'tempText')
+    .attr('font-size', '10')
+    .append('tspan')
+    .text(null);
+  if (fitsWithinNode(tspan, text, textLengthForNodeName)) {
+    return text;
+  }
+  // Computing the size of elipsis is performed with each node. This should be computed once and used everywhere.
+  // TODO: Compute the size of the elipsis once and use it everywhere.
+  const elipsisLength = computeElipsisLength(tspan);
+  let line: string = '';
+  // Calculate how much of the original text to show.
+  // TODO: The folllowing is O(n). We could use a binary search to make this faster: O(log(n)).
+  for (let i = 0; i < text.length; i++) {
+    line += text[i];
+    tspan.text(line);
+    const currentNode = tspan.node();
+    if (currentNode !== null) {
+      const w = currentNode!.getComputedTextLength();
+      if (w >= textLengthForNodeName - elipsisLength) {
+        line = line.slice(0, -1);
+        line += '...';
+        break;
+      }
+    }
+  }
+  tspan.text(null);
+  return line;
```

Comment:
>line [](http://example.com/codeflow?start=9&length=4)

you are not removing the selectAll('.tempText') in this operation. Can it cause some hinderance in computation for subsequent nodes